### DISCUSSION TITLE
How to save 3D Texture?

### DISCUSSION THREAD

**Discussion by jfmilke** [KhronosGroup/KTX-Software#598] [Q&A] [answered]
Hey,
I want to use KTX (version 1) via libktx to save float data as a 3D image. While I figured out how it works for 2D images, I"m lost when it comes to 3D. If someone of you could help me, I would be very glad.

Here is a streamlined code snippet I'm using to save my data as a 3D texture:
```c++
    // Write a KTX(1) Texture
    // In: 'input' vector as float, 'size' struct for dimensions
    ktxTexture1* texture;
    ktxTextureCreateInfo createInfo;
    KTX_error_code result;

    createInfo.glInternalformat = (ktx_uint32_t) gl::GLenum::GL_RGBA32F;
    createInfo.baseWidth = size.x;
    createInfo.baseHeight = size.y;
    createInfo.baseDepth = size.z;
    createInfo.numDimensions = (size.z > 1) ? 3 : 2;
    createInfo.numLevels = 1;
    createInfo.numLayers = 1;
    createInfo.numFaces = 1;
    createInfo.isArray = KTX_FALSE;
    createInfo.generateMipmaps = KTX_FALSE;

    result = ktxTexture1_Create(&createInfo,
      KTX_TEXTURE_CREATE_ALLOC_STORAGE,
      &texture);

    ktx_size_t srcSize = input.size() * sizeof(float);
    uint8_t* src = (uint8_t*)input.data();
    result = ktxTexture_SetImageFromMemory(ktxTexture(texture), 0, 0, 0, src, srcSize);

    ktxTexture_WriteToNamedFile(ktxTexture(texture), "test.ktx");
    ktxTexture_Destroy(ktxTexture(texture));
```

For loading this data I do basically this:
```c++
    // Load a KTX(1) Texture
    ktxTexture1* texture;
    KTX_error_code result;

    result = ktxTexture_CreateFromNamedFile("test.ktx",
      KTX_TEXTURE_CREATE_LOAD_IMAGE_DATA_BIT,
      &ktxTexture(texture));
    
    std::vector<float> out(texture->dataSize / sizeof(float));
    float* fData = reinterpret_cast<float*>(texture->pData);
    out.assign(fData, fData + out.size());

    return out;
```

Help is much appreciated, as for 2D data everything works fine and when it becomes 3D the data is rubbish when I load it again. External tools like AMD's Compressonator also load and display my saved 2D KTX files fine, whereas for my 3D textures they only recognize a Depth of 1 (according to their source info view) and display nothing.

Thanks.

**Edit:** Also when I load my texture using above code `texture->baseDepth` is returned as `1`.

**Solution (thanks @MarkCallow):**
I did not use `ktxTexture_SetImageFromMemory` correctly and hat to use the `faceSlice` parameter for the depth levels of my 3D texture. But I could also omit using `ktxTexture_SetImageFromMemory` and write to `pData` directly.

```c++
    // This worked
    texture->pData = reinterpret_cast<uint8_t*>(input.data());
    ktxTexture_WriteToNamedFile(ktxTexture(texture), path);

    texture->pData = nullptr;
    ktxTexture_Destroy(ktxTexture(texture));
```

```c++
    // And this worked
    uint64_t slice_elements = input.size() / createInfo.baseDepth;
    for (uint32_t z = 0; z < createInfo.baseDepth; z++) {
      ktx_size_t srcSize;
      uint8_t* src;

      srcSize = input.data_size / createInfo.baseDepth;
      src = reinterpret_cast<uint8_t*>(input.data() + z * slice_elements);

      result = ktxTexture_SetImageFromMemory(ktxTexture(texture), 0, 0, z, src, srcSize);

      if (result)
        return false;
    }

    ktxTexture_WriteToNamedFile(ktxTexture(texture), path);
    ktxTexture_Destroy(ktxTexture(texture));
```

**Comment by MarkCallow** [comment, marked as answer]
> I want to use KTX (version 1) via libktx to save float data as a 3D image.

What version of libktx are you using?

I think you will find that `ktxTexture_SetImageFromMemory` is returning `KTX_INVALID_OPERATION` and so not setting any image data. The function is designed to set the data for a single image/depth slice. For 3D textures the third argument is the `slice` to set. The function checks the expected size of an image and, if `srcSize` is different, raises `KTX_INVALID_OPERATION`. For setting the all the images/slices at once, you should use the `pData` pointer as you are in your loader.

I have no idea why `baseDepth` is 1. I cannot reproduce that.

Nothing to do with your problem but why you want to use the KTX 1.0 format?

**Reply by MarkCallow** [reply]
@rovedit please check the value of `texture.baseDepth` immediately after the call to `ktxTexture1_Create` and report back. Also confirm about KTX_INVALID_OPERATION.

**Reply by jfmilke** [reply]
Thanks for your answers!

> What version of libktx are you using?

I'm using the port of package manager [vcpkg](https://github.com/microsoft/vcpkg/tree/master/ports/ktx) which says it's using  4.1.0-rc2 and references commit [b995ac337276648afc3a5beaa6206995d2399bce](https://github.com/KhronosGroup/KTX-Software/commit/b995ac337276648afc3a5beaa6206995d2399bce) from June 15.

>I think you will find that ktxTexture_SetImageFromMemory is returning KTX_INVALID_OPERATION and so not setting any image data. The function is designed to set the data for a single image/depth slice. For 3D textures the third argument is the slice to set. The function checks the expected size of an image and, if srcSize is different, raises KTX_INVALID_OPERATION. For setting the all the images/slices at once, you should use the pData pointer as you are in your loader.

Thanks for this tip, I'll definitely try this tomorrow when I have access to the code again! So I can basically omit `ktxTexture_SetImageFromMemory ` and plainly use `pData` to write my data into the texture? This looks convenient.

> I have no idea why baseDepth is 1. I cannot reproduce that.

Maybe, if `ktxTexture_SetImageFromMemory` did not succeed (I should have checked this..), it didn't write the file to disk at all and I was always loading an old file where I tested 2D textures..

> Nothing to do with your problem but why you want to use the KTX 1.0 format?

Also, I'm using KTX 1.0 format as I'm using OpenGL instead of Vulkan for my project and I don't have the Vulkan enums at hand. It was just a convenient choice to not spend too much tim on finding a solution for this, as I just use KTX as a datacontainer for scientific data (3D vector fields) I want to interpret as image data in another program.

> @rovedit please check the value of texture.baseDepth immediately after the call to ktxTexture1_Create and report back. Also confirm about KTX_INVALID_OPERATION.

I will report tomorrow asap. Sorry for not already answering yesterday. And thanks again for your help.

**Reply by MarkCallow** [reply]
> Maybe, if ktxTexture_SetImageFromMemory did not succeed (I should have checked this..), it didn't write the file to disk at all and I was always loading an old file where I tested 2D textures..

No. It just means the storage allocated by `ktxTexture1_Create` is uninitialized. The only thing that would prevent writing is if `pData` is `NULL`. Check for errors after each libktx call.

**Reply by jfmilke** [reply]
It works! Thank you very much.

`ktxTexture_SetImageFromMemory` indeed returned a `KTX_INVALID_OPERATION`. Using the `faceSlice` parameter solved the issue, as did writing to `pData` directly. Though, I would not have guessed that this could be the right parameter, as I thought it would only apply to cubemap (and thus for its faces).

To summarize:
```c++
    // This worked
    texture->pData = reinterpret_cast<uint8_t*>(input.data());
    ktxTexture_WriteToNamedFile(ktxTexture(texture), path);

    texture->pData = nullptr;
    ktxTexture_Destroy(ktxTexture(texture));
```

```c++
    // And this worked
    uint64_t slice_elements = input.size() / createInfo.baseDepth;
    for (uint32_t z = 0; z < createInfo.baseDepth; z++) {
      ktx_size_t srcSize;
      uint8_t* src;

      srcSize = input.data_size / createInfo.baseDepth;
      src = reinterpret_cast<uint8_t*>(input.data() + z * slice_elements);

      result = ktxTexture_SetImageFromMemory(ktxTexture(texture), 0, 0, z, src, srcSize);

      if (result)
        return false;
    }

    ktxTexture_WriteToNamedFile(ktxTexture(texture), path);
    ktxTexture_Destroy(ktxTexture(texture));
```

**Reply by MarkCallow** [reply]
> I would not have guessed that this could be the right parameter

The parameter is called `faceSlice` and the documentation says "cube map face or depth slice of the image to set". How could I make it clearer?

I'm glad you got it working.

**Reply by jfmilke** [reply]
Yeah, sorry, I totally overlooked that. My fault. I think I just got hung up at the face-part.
